stetho/src/main/java/com/facebook/stetho/inspector/network/AbstractAsyncPrettyPrinterFactory.java

```diff
@@ -0,0 +1,85 @@
+package com.facebook.stetho.inspector.network;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import com.facebook.stetho.common.ExceptionUtil;
+import com.facebook.stetho.common.Util;
+import com.facebook.stetho.inspector.network.AsyncPrettyPrinter;
+import com.facebook.stetho.inspector.network.AsyncPrettyPrinterFactory;
+
+import javax.annotation.Nullable;
+
+/**
+ * Abstract class for pretty printer factory that asynchronously downloads schema needed for
+ * pretty printing the payload
+ */
+public abstract class AbstractAsyncPrettyPrinterFactory implements AsyncPrettyPrinterFactory {
+
+  private static ExecutorService sExecutorService = Executors.newFixedThreadPool(1);
+
+  @Override
+  public AsyncPrettyPrinter getInstance(String headerName, String headerValue) {
+    try {
+      String uri = parseHeaderValueToUri(headerName, headerValue);
+      final Future<Response> response = sExecutorService.submit(new Request(new URL(uri)));
+      return new AsyncPrettyPrinter() {
+        public void printTo(PrintWriter output, InputStream payload)
+            throws IOException {
+          try {
+            Response schemaResponse = Util.getUninterruptibly(response);
+            doPrint(output, payload, schemaResponse.getBody());
+          } catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+            ExceptionUtil.propagateIfInstanceOf(cause, IOException.class);
+            throw ExceptionUtil.propagate(cause);
+          }
+        }
+      };
+    } catch (MalformedURLException e) {
```

Comment:
Zen of try/catch management in Java, tip number 192: use isolate annoying checked exceptions into their own simple static functions to improved readability:

```java
@Nullable
private static URL parseURL(String uri) {
  try {
    return new URL(url);
  } catch (...) {
    return null;
  }
}
```